File: Python/Strings/1024 - Criptografia.py
"""
Autor: Pierre Vieira
Data da submissão: 27/01/2019 01:32:11
"""
casos_teste = int(input())
while casos_teste != 0:
    string = input()
    texto = list(string)
    if not(1 <= len(string) <= 1000):
        break
    for c in range(0, len(string)):
        if texto[c].isalpha():
            texto[c] = chr(ord(texto[c]) + 3)
    ultimo_caractere = len(string) -1
    for c in range(0, len(string)//2):
        aux = texto[c]
        texto[c] = texto[ultimo_caractere]
        texto[ultimo_caractere] = aux
        ultimo_caractere -= 1
    for c in range(len(string)//2, len(string)):
        texto[c] = chr(ord(texto[c]) - 1)
    print(''.join(texto))
    casos_teste -= 1

File: Kotlin/Strings/ 1024 - Criptografia.kt
/*
* Autor: Pierre Vieira
* Data de submissão: 06/03/2023 11:13:32
*/

private fun thirdStep(string: String) : String {
    var mappedString = ""
    val limitedIndex = string.length / 2
    string.forEachIndexed { index: Int, char: Char ->
        mappedString += if (index >= limitedIndex) char.dislocate(-1) else char
    }
    return mappedString
}

private fun secondStep(string: String) : String = string.reversed()

private fun firstStep(string: String) : String {
    var mappedString = ""
    string.forEach {
        mappedString += if (it.isLowerCase() || it.isUpperCase()) it.dislocate(3).toString() else it
    }
    return mappedString
}

private fun Char.dislocate(positions: Int): Char = (toInt() + positions).toChar()

fun main(args: Array<String>) {
    for (i in 0 until readLine()!!.toInt()) {
        val input = readLine()!!
        val mappedInput = thirdStep(secondStep(firstStep(input)))
        println(mappedInput)
    }
}